internal/internal_deployment_client.go

```diff
@@ -0,0 +1,255 @@
+		CreateTime: deployment.GetCreateTime().AsTime(),
+		IsCurrent:  deployment.GetIsCurrent(),
+	}
+}
+
+func deploymentTaskQueuesInfoFromProto(tqsInfo []*deployment.DeploymentInfo_TaskQueueInfo) []DeploymentTaskQueueInfo {
+	result := []DeploymentTaskQueueInfo{}
+	for _, info := range tqsInfo {
+		result = append(result, DeploymentTaskQueueInfo{
+			Name:            info.GetName(),
+			Type:            TaskQueueType(info.GetType()),
+			FirstPollerTime: info.GetFirstPollerTime().AsTime(),
+		})
+	}
+	return result
+}
+
+func deploymentInfoFromProto(deploymentInfo *deployment.DeploymentInfo) DeploymentInfo {
+	return DeploymentInfo{
+		Deployment:     deploymentFromProto(deploymentInfo.GetDeployment()),
+		CreateTime:     deploymentInfo.GetCreateTime().AsTime(),
+		IsCurrent:      deploymentInfo.GetIsCurrent(),
+		TaskQueuesInfo: deploymentTaskQueuesInfoFromProto(deploymentInfo.GetTaskQueueInfos()),
+		Metadata:       deploymentInfo.GetMetadata(),
+	}
+}
+
+func deploymentReachabilityInfoFromProto(response *workflowservice.GetDeploymentReachabilityResponse) DeploymentReachabilityInfo {
+	return DeploymentReachabilityInfo{
+		DeploymentInfo: deploymentInfoFromProto(response.GetDeploymentInfo()),
+		Reachability:   DeploymentReachability(response.GetReachability()),
+		LastUpdateTime: response.GetLastUpdateTime().AsTime(),
+	}
+}
+
+func deploymentMetadataUpdateToProto(dc converter.DataConverter, update DeploymentMetadataUpdate) *deployment.UpdateDeploymentMetadata {
+	upsertEntries := make(map[string]*common.Payload)
+
+	for k, v := range update.UpsertEntries {
+		if enc, ok := v.(*common.Payload); ok {
+			upsertEntries[k] = enc
+		} else {
+			dataConverter := dc
+			if dataConverter == nil {
+				dataConverter = converter.GetDefaultDataConverter()
+			}
+			metadataBytes, err := dataConverter.ToPayload(v)
+			if err != nil {
+				panic(fmt.Sprintf("encode deployment metadata error: %v", err.Error()))
+			}
+			upsertEntries[k] = metadataBytes
+		}
+	}
+
+	return &deployment.UpdateDeploymentMetadata{
+		UpsertEntries: upsertEntries,
+		RemoveEntries: update.RemoveEntries,
+	}
+}
+
+func (dc *deploymentClient) List(ctx context.Context, options DeploymentListOptions) (DeploymentListIterator, error) {
+	paginate := func(nextToken []byte) (*workflowservice.ListDeploymentsResponse, error) {
+		grpcCtx, cancel := newGRPCContext(ctx, defaultGrpcRetryParameters(ctx))
+		defer cancel()
+		request := &workflowservice.ListDeploymentsRequest{
+			Namespace:     dc.workflowClient.namespace,
+			PageSize:      int32(options.PageSize),
+			NextPageToken: nextToken,
+			SeriesName:    options.SeriesName,
+		}
+
+		return dc.workflowClient.workflowService.ListDeployments(grpcCtx, request)
+	}
+
+	return &deploymentListIteratorImpl{
+		paginate: paginate,
+	}, nil
+}
+
+func (dc *deploymentClient) Describe(ctx context.Context, options DeploymentDescribeOptions) (DeploymentInfo, error) {
+	if err := dc.workflowClient.ensureInitialized(ctx); err != nil {
```

Comment:
Should add this to the top of `List` too

Reply:
Good point, thanks. I added it inside `paginate()`, `List` just returns the iterator struct...